### PR TITLE
Rx and Ry distances for sources in source modelling

### PR DESCRIPTION
**Pull request by lispandfound**
Calculates:

1. Planar rx/ry between one plane and many points,
2. Fault rx/ry between one fault and many points,
3. Scenario rx/ry between several faults and many points.

I have introduced a new `gc2_distances` module for the calculation, and several extensive tests for every part of the process.

Here is "proof by picture" that the calculations appear to work using a dummy pair of faults.

rx_calculations

I will dogfood this for my own hazard calculations and make any further fixes.

### CODE REVIEW

**Comment by lispandfound** [comment]
This PR also contains the three line changes from #57 because Fiona was once again killing the CI pipeline

File: source_modelling/gc2_distances.py [hunk 1]
```diff
@@ -0,0 +1,405 @@
...
+            out=np.full_like(w, np.nan),
+        ) - np.divide(1.0, ry, where=(ry != 0), out=np.full_like(w, np.nan))
+
+        w = np.where(mask_zero, special_case, w)
+
+    return w
+
+
+@njit([float64[:](float64[:], uint64[:]), float64[:](float64[:], int64[:])], cache=True)
+def cumulative_reduction(data: np.ndarray, indices: np.ndarray) -> np.ndarray:
```
Comment (claudio525) [inline source_modelling/gc2_distances.py]: I assume we need this because `np.cumulative_sum` is needed inside numba optimized function and they don't support it at this stage? If so, that might be worth mentioning in the docstring.

**Review by AndrewRidden-Harper — DISMISSED** [review]
Looks great. Very comprehensive tests

File: source_modelling/gc2_distances.py (hunk 1 above)
Comment (claudio525) [reply source_modelling/gc2_distances.py]: Actually, it doesn't look to be used inside a numba function, so I'm confused as to why we can't just use `np.cumulative_sum`?

**Review by claudio525 — COMMENTED** [review]
This looks great, nice work! Will be great to have this as part of `source_modelling`.

File: source_modelling/gc2_distances.py (hunk 1 above)
Comment (lispandfound) [reply source_modelling/gc2_distances.py]: @claudio525 because the cumulative sum adds up across the whole array, but I want to accumulate segment lengths between indices. The way this code stays fast and vectorised is by flattening the ragged structure of the faults so that all the trace lengths sit in memory in a 1d numpy array, accompanied by indices that track where segments start and end. The cumulative sum ignores that structure and sums trace lengths across different segments, where as this numba kernel resets the sum between individual segments.

Image

**Review by claudio525 — DISMISSED** [review]

**Comment by lispandfound** [comment]
Summary of changes since last review:

1. I ran a ruff format over the repo. Ignore all changes except the `gc2_distances.py`, `sources.py` module, the `test_sources.py` and `test_rx_ry.py` test code.
2. Fiona no longer prevents tests running.
3. I had to handle an extra edge case for multi trace rx-ry where the a point lies precisely on one segment. I forgot to handle this case, but it explains in the referenced paper that the generalised rx, ry distance metrics should collapse to the rx, ry of the segment the point is contained in. I've attached a picture of the three weight equations depending on the relation between the query point and a given trace. The case marked `*` is the case that I forgot.
image

**Review by claudio525 — DISMISSED** [review]

**Review by AndrewRidden-Harper — DISMISSED** [review]
